File: app/src/main/java/com/github/panda_io/micro_panda/builder/llvm/ir/global.go
package ir

import (
	"fmt"
	"io"

	"github.com/panda-io/micro-panda/target/llvm/ir/constant"
	"github.com/panda-io/micro-panda/target/llvm/ir/ir"
	"github.com/panda-io/micro-panda/target/llvm/ir/ir_types"
)

// constant
type Global struct {
	ir.GlobalIdent
	Immutable   bool
	ContentType ir.Type
	Init        constant.Constant
	Typ         *ir_types.PointerType
}

func NewGlobal(name string, contentType ir.Type) *Global {
	global := &Global{ContentType: contentType}
	global.SetName(name)
	global.Type()
	return global
}

func NewGlobalDef(name string, init constant.Constant) *Global {
	global := &Global{ContentType: init.Type(), Init: init}
	global.SetName(name)
	global.Type()
	return global
}

func (g *Global) String() string {
	return fmt.Sprintf("%s %s", g.Type().String(), g.Ident())
}

func (g *Global) Type() ir.Type {
	if g.Typ == nil {
		g.Typ = ir_types.NewPointerType(g.ContentType)
	}
	return g.Typ
}

func (g *Global) WriteIR(w io.Writer) error {
	declaration := "global"
	if g.Immutable {
		declaration = "constant"
	}
	_, err := fmt.Fprintf(w, "%s = %s %s", g.Ident(), declaration, g.ContentType.String())
	if err != nil {
		return err
	}
	if g.Init != nil {
		_, err = fmt.Fprintf(w, " %s", g.Init.Ident())
	}
	return err
}

func (g *Global) IsConstant() {}

File: app/src/main/java/com/github/panda_io/micro_panda/builder/llvm/ir/Global.java
package com.github.panda_io.micro_panda.builder.llvm.ir;

import com.github.panda_io.micro_panda.builder.llvm.ir.constant.Constant;
import com.github.panda_io.micro_panda.builder.llvm.ir.type.Pointer;
import com.github.panda_io.micro_panda.builder.llvm.ir.type.Type;

public class Global extends Constant {
    Identifier _identifier;
    boolean immutable;
    Type contentType;
    Constant initializer;
    Type type;

    public Global(String name, Constant initializer) {
        this._identifier = new Identifier(true);
        this._identifier.setName(name);
        this.initializer = initializer;
        this.contentType = this.initializer.getType();
        this.getType();
    }

    public String string() {
        return String.format("%s %s", this.type.string(), this.identifier());
    }

    public Type getType() {
        if (this.type == null) {
            this.type = new Pointer(this.contentType);
        }
        return this.type;
    }

    public String identifier() {
        return this._identifier.identifier();
    }

    public void writeIR(StringBuilder builder) {
        String declaration = "global";
        if (this.immutable) {
            declaration = "constant";
        }
        builder.append(String.format("%s = %s %s %s", this.identifier(), declaration, this.contentType.string(),
                this.initializer.identifier()));
    }
}